src/Sulu/Bundle/DocumentManagerBundle/Reference/Subscriber/DocumentReferenceSubscriber.php

```diff
@@ -0,0 +1,210 @@
+            Events::PERSIST => 'onPersist',
+            Events::REMOVE => 'onRemove',
+            Events::REMOVE_LOCALE => 'onRemoveLocale',
+            Events::CLEAR => 'onClear',
+            Events::FLUSH => 'onFlush'
+        ];
+    }
+
+    public function onPublish(PublishEvent $event): void
+    {
+        $document = $event->getDocument();
+        $locale = $event->getLocale();
+
+        if (!$document instanceof StructureBehavior
+            || !$document instanceof TitleBehavior
+            || !$document instanceof UuidBehavior
+        ) {
+            return;
+        }
+
+        $this->publishDocuments[] = [
+            'document' => $document,
+            'locale' => $locale,
+        ];
+    }
+
+    public function onPersist(PersistEvent $event): void
+    {
+        $document = $event->getDocument();
+        $locale = $event->getLocale();
+
+        if (!$document instanceof StructureBehavior
+            || !$document instanceof TitleBehavior
+            || !$document instanceof UuidBehavior
+        ) {
+            return;
+        }
+
+        $this->persistDocuments[] = [
+            'document' => $document,
+            'locale' => $locale,
+        ];
+    }
+
+    public function onRemove(RemoveEvent $event): void
+    {
+        $document = $event->getDocument();
+
+        if (!$document instanceof StructureBehavior
+            || !$document instanceof UuidBehavior
+        ) {
+            return;
+        }
+
+        $this->removeDocuments[] = [
+            'document' => $document,
+            'locale' => null,
+        ];
+    }
+
+    public function onRemoveLocale(RemoveLocaleEvent $event): void
+    {
+        $document = $event->getDocument();
+
+        if (!$document instanceof StructureBehavior
+            || !$document instanceof UuidBehavior
+        ) {
+            return;
+        }
+
+        $this->removeDocuments[] = [
+            'document' => $document,
+            'locale' => $event->getLocale(),
+        ];
+    }
+
+    public function onClear(ClearEvent $event): void
+    {
+        $this->reset();
+    }
+
+    public function onFlush(FlushEvent $event): void
+    {
+        foreach ($this->persistDocuments as $documentData) {
+            $this->getProvider($documentData['document'])?->updateReferences(
+                $documentData['document'],
+                $documentData['locale'],
+            );
+        }
+
+        foreach ($this->publishDocuments as $documentData) {
+            $this->getProvider($documentData['document'])?->updateReferences(
+                $documentData['document'],
+                $documentData['locale'],
+            );
+        }
+
+        foreach ($this->removeDocuments as $documentData) {
+            $this->getProvider($documentData['document'])?->removeReferences(
+                $documentData['document'],
+                $documentData['locale'],
+            );
+        }
+
+        $this->reset();
```

Comment:
@alexander-schranz Currently we get duplicated entries, the problem is, that we are clearing twice in the same transaction (once for the `persistDocuments` and once for the `publishDocuments`), most likely the delete flush is then only executed once in the transaction, but the insert statements from both (persist and publish) are executed, resulting in duplicated entries when you press `Save & Publish`